pyccel/complexity/basic.py

```diff
@@ -25,16 +39,242 @@ def __init__(self, filename_or_text):
+        if mode and mode != 'simple':
+            raise ValueError('not valid value for mode')
+        self._mode = mode
+        # ...
+
+        # ...
+        costs = OrderedDict()
+
+        # ... first we treat declared functions
+        if self.functions:
+            for fname, d in self.functions.items():
+                expr =  self._cost(d)
+
+                if simplify:
+                    expr = sp_simplify(expr)
+
+                if bigo:
+                    for n in bigo:
+                        expr = LT(expr, Symbol(n))
+
+                costs[fname] = expr
+
+        self._costs.update(costs)
+        # ...
+
+        # ... then we compute the complexity for the main program
+        expr = self._cost(self.ast.program)
+
+        if simplify:
+            expr = sp_simplify(expr)
+
+        if bigo:
+            for n in bigo:
+                expr = LT(expr, Symbol(n))
+        # ...
+
+        return expr
+
+    def _cost(self, expr, **settings):
+        if expr is None:
+            return 0
+
+#        print('>>> ', expr, type(expr))
+
+        classes = type(expr).__mro__
+        for cls in classes:
+            method = '_cost_' + cls.__name__
+            if hasattr(self, method):
+                obj = getattr(self, method)(expr, **settings)
+                return obj
+        raise NotImplementedError('{} not available for {}'.format(method, type(expr)))
+
+    def _cost_Program(self, expr, **settings):
+        return self._cost(expr.body, **settings)
+
+    def _cost_CodeBlock(self, expr, **settings):
+        return sum(self._cost(i, **settings) for i in expr.body)
+
+    def _cost_Comment(self, expr, **settings):
+        return 0
+
+    def _cost_EmptyNode(self, expr, **settings):
+        return 0
+
+    def _cost_Tuple(self, expr, **settings):
+        return sum(self._cost(i, **settings) for i in expr)
+
+    def _cost_list(self, expr, **settings):
+        return sum(self._cost(i, **settings) for i in expr)
+
+    def _cost_tuple(self, expr, **settings):
+        return sum(self._cost(i, **settings) for i in expr)
+
+    def _cost_LiteralInteger(self, expr, **settings):
+        return 0
+
+    def _cost_LiteralFloat(self, expr, **settings):
+        return 0
+
+    def _cost_For(self, expr, **settings):
+        ops = sum(self._cost(i, **settings) for i in expr.body.body)
+
+        # ...
+        i = expr.target
+        i = pyccel_to_sympy(i, self._symbol_map, self._used_names)
+
+        iter_range = expr.iterable.get_range()
+        b = iter_range.start
+        b = pyccel_to_sympy(b, self._symbol_map, self._used_names)
+
+        e = iter_range.stop
+        e = pyccel_to_sympy(e, self._symbol_map, self._used_names)
+
+        step = iter_range.step
+        step = pyccel_to_sympy(step, self._symbol_map, self._used_names)
+        # ...
+
+        end = ((e-b) / step)-1
+        # translate so that we always start from 0
+        return summation(ops, (i, 0, end))
```

Comment:
Translating so that the index always starts from 0 leads to wrong results if loops inside loops depend on the index. E.g:
```
def test(degree : int):
    a = 0
    for j in range(1, degree+1):
        for r in range(j):
            a += 1
return a
```